src/main/preload/externalAPI.ts

```diff
@@ -0,0 +1,470 @@
+        return result && result.length > 0;
+    });
+    ipcRenderer.send(UNREAD_RESULT, isUnread);
+};
+
+let sessionExpired: boolean;
+const getUnreadCount = () => {
+    // LHS not found => Log out => Count should be 0, but session may be expired.
+    let isExpired;
+    if (document.getElementById('sidebar-left') === null) {
+        const extraParam = (new URLSearchParams(window.location.search)).get('extra');
+        isExpired = extraParam === 'expired';
+    } else {
+        isExpired = false;
+    }
+    if (isExpired !== sessionExpired) {
+        sessionExpired = isExpired;
+        ipcRenderer.send(SESSION_EXPIRED, sessionExpired);
+    }
+};
+
+/**
+ * Legacy message passing code - can be running alongside the new API stuff
+ */
+
+// Disabling no-explicit-any for this legacy code
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+window.addEventListener('message', ({origin, data = {}}: {origin?: string; data?: {type?: string; message?: any}} = {}) => {
+    const {type, message = {}} = data;
+    if (origin !== window.location.origin) {
+        return;
+    }
+    switch (type) {
+    case 'webapp-ready':
+    case 'get-app-version': {
+        // register with the webapp to enable custom integration functionality
+        ipcRenderer.invoke(GET_APP_INFO).then((info) => {
+            console.log(`registering ${info.name} v${info.version} with the server`);
+            window.postMessage(
+                {
+                    type: 'register-desktop',
+                    message: info,
+                },
+                window.location.origin || '*',
+            );
+        });
+        break;
+    }
+    case 'dispatch-notification': {
+        const {title, body, channel, teamId, url, silent, data: messageData} = message;
+        channels.set(channel.id, channel);
+        ipcRenderer.send(NOTIFY_MENTION, title, body, channel.id, teamId, url, silent, messageData.soundName);
+        break;
+    }
+    case BROWSER_HISTORY_PUSH: {
+        const {path} = message as {path: string};
+        ipcRenderer.send(BROWSER_HISTORY_PUSH, path);
+        break;
+    }
+    case 'history-button': {
+        ipcRenderer.invoke(REQUEST_BROWSER_HISTORY_STATUS).then(sendHistoryButtonReturn);
+        break;
+    }
+    case CALLS_LINK_CLICK: {
+        ipcRenderer.send(CALLS_LINK_CLICK, message.link);
+        break;
+    }
+    case GET_DESKTOP_SOURCES: {
+        ipcRenderer.invoke(GET_DESKTOP_SOURCES, message).then(sendDesktopSourcesResult);
+        break;
+    }
+    case CALLS_WIDGET_SHARE_SCREEN: {
+        ipcRenderer.send(CALLS_WIDGET_SHARE_SCREEN, message.sourceID, message.withAudio);
+        break;
+    }
+    case CALLS_JOIN_CALL: {
+        ipcRenderer.invoke(CALLS_JOIN_CALL, message).then(sendCallsJoinedCall);
+        break;
+    }
+    case CALLS_JOINED_CALL: {
+        ipcRenderer.send(CALLS_JOINED_CALL, message.callID);
+        break;
+    }
```

Comment:
So this is where we would want versioned methods. First off, we will not be modifying the legacy code anymore so we wouldn't be adding here, we would be modifying the above `contextBridge` code.

In this case if we're just adding another parameter, we should be able to just tack it on to the existing call as optional, and then we check for it when we need to consume it.

Reply:
@devinbinnie Not sure I understand how versioned methods would solve the problem of having to update the Desktop side every time we want to extend the object we are passing with some new fields. The way it was working before was convenient since the data to forward was completely opaque to the desktop which made it easy to extend without requiring a new release and all that it entails.

But regardless of the above, we should update both the new and legacy code now or it will break some things Calls wise. I can take care of that if you prefer, it's about passing an additional `sessionID` property coming through the `CALLS_JOINED_CALL` event.

Reply:
> Not sure I understand how versioned methods would solve the problem of having to update the Desktop side every time we want to extend the object we are passing with some new fields.

I think we're missing something else, because if you plan on having the widget or the Desktop App send new fields back for the other side to listen, wouldn't you need to update both sides anyways? I guess in your case there might be some strangeness if you're just using the Desktop App to forward between them, but if that's the case then I think we actually just have a design problem here wrt. the Calls Widget. Generally the API should be used to facilitate communication between the Desktop App and the Web App, or the Desktop App and a plugin. If we're using it only to relay, I think we're actually misusing it. In your case here, we might want to come up with a different solution.

> But regardless of the above, we should update both the new and legacy code now or it will break some things Calls wise.

Is there something I omitted in the legacy code that was there before? Or are you adding something new now? If it's something new, we should not be updating the legacy code since I don't want to extend support on it. I think if we're adding something we should accelerate the API as opposed to continuing to support the old message passing. 3/5 on this.

Reply:
> If we're using it only to relay, I think we're actually misusing it. In your case here, we might want to come up with a different solution.

One solution I could suggest, though security might want to get involved here is that you could set up a relay listener for the calls widget specifically that allows for arbitrary data to be passed through.

This could be secure as long as we do it properly by:
1) Ensuring only serializable data can flow through the endpoint, which is something Electron already guarantees, so that's good :)
2) Making sure we check the origin and destination of each request to make sure that we don't send things to strange places
3) Perform some kind of validation on the data received and sent (this could be done at either end if we want to avoid changing the Desktop App)

Honestly though, I kinda see the above solution as a bit of a hack and again a bit of a misuse of this API. The idea behind the API is to be deliberate, strict, and as limited as necessary, so exposing something like this feels like it goes against the design principles.

Happy to talk it over and help out with the solution :)

Reply:
Okay what you are suggesting would be taking it even further from strictness. I sort of liked the balance of having a well defined event but the data contained could be extended. 

As an example, the `CALLS_ERROR`:

```ts
export type CallsErrorMessage = {
    err: string;
    callID?: string;
    errMsg?: string;
}
```

An error can happen on the widget window but the modal is displayed in the webapp (main view) so that's why we relay it through desktop even though there's nothing that desktop should be really aware of or do to the data. Now, if in the future we want to add a new field, say an error code, we'd have to send a PR here, wait for a new Desktop release and likely make the plugin backwards compatible as well (handle both desktop versions).

If you think we should be strict throughout then that's fine, just want to make sure this problem is highlighted and we accept the trade-offs.

/cc @cpoile 


> Is there something I omitted in the legacy code that was there before? Or are you adding something new now? If it's something new, we should not be updating the legacy code since I don't want to extend support on it. I think if we're adding something we should accelerate the API as opposed to continuing to support the old message passing. 3/5 on this.

You omitted something because you weren't aware of it due to the reason above. We sneaked in a new `sessionID` field without changing anything on this side. But Calls already is released and it's working so if we don't update the code now, this release will break for any Calls version not implementing the new API. It's a minor change which I think we should add before locking the legacy API given the affected functionality has been released already,

Reply:
> Okay what you are suggesting would be taking it even further from strictness. I sort of liked the balance of having a well defined event but the data contained could be extended.

I agree here, but I think the fact that the plan was to allow arbitrary fields to change would also make sure this wasn't deliberate either. This way we would have a deliberate endpoint to say "send some data from the calls widget to the Web App". It would be "less strict" but would accomplish what you need.

To be honest I don't really like that solution, as said it's kind of a misuse of the API.

> You omitted something because you weren't aware of it due to the reason above. We sneaked in a new `sessionID` field without changing anything on this side. But Calls already is released and it's working so if we don't update the code now, this release will break for any Calls version not implementing the new API. It's a minor change which I think we should add before locking the legacy API given the affected functionality has been released already.

Right, so this is the exact problem I highlighted above. We are making use of the Desktop App to send something through it that it has no awareness of, or as you put it, "sneaked in". This to me is a big red flag in terms of design, and a bit of a security concern. We shouldn't be expecting that outside sources should have any say as to what goes through the Desktop App. The above endpoint was kind of my compromise which would allow a deliberate endpoint that does what you need it to.

But I think we need to change how this works fundamentally, which might mean that the calls widget becomes part of the Desktop App, we find an alternate way to have the calls widget (ie. external code) talk to different external code running in a different browser instance, or we just accept that when we change the Calls Widget API that we also change the Desktop App.

I can definitely see the benefit of letting the Calls Widget do whatever it wants in terms of getting rapid changes out fast, and that's great, but I think we went about this wrong due to the way was designed in the first place, which is kind of what I'm trying to fix with this change.

As for the specific field above, we can add that one now but I want to set the expectation that we won't be doing any further changes to the legacy API going forward, and that we'll need to come up with a way forward.

Reply:
> This to me is a big red flag in terms of design, and a bit of a security concern. We shouldn't be expecting that outside sources should have any say as to what goes through the Desktop App.

Fair enough, I am not going to argue against that. We'll keep it strict from now onward then.

> or we just accept that when we change the Calls Widget API that we also change the Desktop App.

I fear this is the only way forward for the time being as anything else would likely require some unplanned effort. That said, I am very interested in chatting more on this as I may start to agree that the standalone widget logic living on this side is indeed the way to go. May be something good to track for Calls v1 which we are planning sometime in Q1 next year.

> As for the specific field above, we can add that one now but I want to set the expectation that we won't be doing any further changes to the legacy API going forward, and that we'll need to come up with a way forward.

Cool. This is just to ensure no breaking changes are made due to this. I don't expect any further addition to the legacy stuff. Let me know if you need any help with that.

Reply:
> I fear this is the only way forward for the time being as anything else would likely require some unplanned effort. That said, I am very interested in chatting more on this as I may start to agree that the standalone widget logic living on this side is indeed the way to go. May be something good to track for Calls v1 which we are planning sometime in Q1 next year.

I'm kind of feeling the same as well. It would be good to have the Desktop App control more of it's own specific code where it can. We can definitely chat more on the specifics going forward.

Also, yeah I can add the `sessionID` bit, will message you separately to make sure I have it right :P